Cloudant Sync Core Data Incremental Store

jimix opened the pull request.

jimix committed:
Phase 1 of NSIncrementalStore work

We can fetch, store, update and delete
BUG:
  - if we make a change in a relation, the next time we run, nothing loads?
TODO:
  - Come up with a better way for uniqueID
  - queries, and others..
  - relations are straight forward and simple
  - all _very_ naive and inefficient, I'm sure
  - Sync with backend

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Create a dot graph, helpful for debugging.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
SUCK! Caches, get out of sync over runs somehow, so we delete them.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Add support for sort descriptors

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Change encoding so that values for predicate and sort are off the top level.

This requires us to split metatdata from data, it's not so bad.
However, The CDT docs do state that there is a way to control where
the value comes from the index but it is poorly explained (at least to
me) and will likely slow stuff down.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
UNTESTED attempt at predicates

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
First IncrementalStore Unit Test

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
More fields for Entry for testing

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
More Code Selectors

1. Verify backingstore document survives JSON serialization
2. Update the DOT graph on every save

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Deal with JSON double limitations.

JSON cannot handle the full range of double so we store
two values:
1. `long long` "image" so we can store accurately
  > This could be a problematic when replicating to other arches
2. A JSON range value so predicates and sort order can work

Right now the JSON conversion is to just make it a float.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Sort descriptors need to know the index type

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Add fetch constraints

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Nits!

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Support all available predicates

Cause an exception when others are used
Includes Unit Tests

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Static method to find store coordinators that are of our class.

Thought I would need it. but I don't yet.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Replication: part 1: create the push and pull replicator objects

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Enable Logging

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
oops cleanup

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Documentation for Core Data and Incremental Store

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Add Unit test for expressionForEvaluatedObject using NSInPredicateOperatorType

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
NSLog messages should use logging facility

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixup: prefix predicate code switch

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
unittest: fixup: super teardown last

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Add convenience function where direction is an argument.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixup: unitTest: proper attribution

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixup: spelling

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Add Replication Acceptance Test for Core Data

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Logging: add more logging options and prefix our own messages

jimix committed:
Resolves #8: Document Portability Issues and Solutions

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Resolves #7: Add class method to declare mime-type of transformed objects

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Closes #9: Switch to a Dictionary based encoding rather than an array base.

Ultimately, this helps future proof the objects and make the documents
more descriptive.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Test for NaN for NSDecimal

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

Conflicts:
	Tests/Tests/IncrementalStore.m

jimix committed:
Closes #3: Allow NSDecimalNumber to be usable in a predicate

Also stores a binary image so it can be used for accurate restore

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Fixes #10: Stop demoting doubles to float to make JSON parser happy.

NSDecimalNumber "description" is the closest thing we will get to an
arbitrary precision number in JSON, so lets use it.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Nits and fix print warnings in unit tests

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
RA: Check the number of documents in the remote replicant

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Fixes #1: oops, I pushed when I should have pulled

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixme: Add Replication Acceptance Test for Core Data

jimix committed:
fixme: Add Replication Acceptance Test for Core Data: actually fetch pulled replcant in to core data

jimix committed:
fixme: get replication status from rep, not pusher

jimix committed:
Fixes #2: NSFetchedResultsController must be purged

It is unfortunate that we do not know the name of the cache, so we
blow them all away.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Add test for deduplication

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Fixes #11: Limited support for NSDictionaryResultType fetches

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
more oops cleanup

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Allow oops() from C functions

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
encode hashes using hexstrings rather than base64

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixup: formatting

jimix committed:
Fixes #6: Move Meta data into the metadata object

All fixed attributes, mainly type, that are noted as metadata are now
in the metadata document.

We also store object hashes so we can compare them in the future.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Closes #14: Remove revIDFromDocID

It was useless anyway

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Add a descriptive label to the document IDs

So they can be identified easily out of the application.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixup: remove redundant uses of allKeys

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Fixes #12: use attachements for binary and transformable object

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixup: TypeName is really PropertyName

jimix committed:
Fixes #15: Remove "per-relation" destination information and get it from metadata

It is constant after all

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
get xform class name from metatdata

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
MetaData changes inspired by JSON schema creation and audit

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Document JSON Schema for the Mata Data Document

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Closes #13: Document JSON Schema for Entities

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixup: tabify metadata schema

jimix committed:
fixup: tabs in coredata.md

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixup: comments are not legal in JSON so we view the schemas using "javascript"

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Error enum should be declared in header

The app needs to see them.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixup: OSX warns about undeclared selector

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixup: abstract out CD metadata encoding

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
closes #16: remove `CDTISType`

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Switch to using the Core Data Model Editor

This should make it easier to test object versions and migrations.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixup: rename CoreDataEntry to CDEv1.0

jimix committed:
Fix #20: Deal with Transformable objects with no transformer class

Transformable objects should use NSKeyedArchiver it a transformer
class is not defined

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Closes #19: Transform the given database name into a name acceptable to CouchDB

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
the userInfo dictionaries are not seen by the backing store

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

mkistler commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2798 @@
+
+        case NSStringAttributeType:
+            self.typeName = kCDTISStringAttributeType;
+            break;
```
typo: tha->the

mkistler commented on Classes/common/CDTIncrementalStore.h:
```diff
@@ -0,0 +1,121 @@
+ *
+ *  @return NSString
+ */
++ (NSString *)type;
```
Curious that this is a class method and not an instance method.  Same comment for localDir.

mkistler replied:
After looking at the implementation, I see that a class method is correct.  But I wonder if the description is really accurate.  Perhaps it should be "the string that should be used to register this incremental store"?

jimix replied:
NACK,
This document snippet is for the application programmer so they can identify this IS to Core Data when they add the store.
The act of "registering" the IS is internal to our code.

mkistler commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2798 @@
+@implementation CDTISProperty
+
+/**
+ * Defines tha attribute meta data that is stored in the object.
```
Another tha.  Might want to do a global search on ' tha '.

mkistler replied:
Typo: tha->the

jimix replied:
ACK,
I better check my personal dictionary, I may have added it as a word :)

mkistler commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2798 @@
+ */
+- (CDTISProperty *)initWithAttribute:(NSAttributeDescription *)attribute
+{
+    self.isRelationship = NO;
```
I think initializers should follow the "initializer pattern", which is something like:

self = [super init];
if (self) {
    // Perform custom init for this class
}
return self;

Is there some reason you did not do that here?

mkistler replied:
Another style issue with initializers: I think it is best to avoid using 'self.' when referencing properties in the initializer, since this requires that all getters and setters behave properly even on partially initialized objects.  It is better to set the underlying instance variables directly in the initializer to avoid this dependency.

jimix replied:
ACK, on `[super init]`and "instance variables"

mkistler commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2798 @@
+    return [NSDictionary dictionaryWithDictionary:dic];
+}
+
+- (NSInteger)typeCodeFromName:(NSString *)name
```
Since this method does not reference any properties of the object, it is effectively a class method, so for clarity I would declare it as such.

jimix replied:
ACK
My preference is to actually make convert methods that do not reference to object as a static C function.  This would be consistent with the rest of the code.  I personally find calling class methods cumbersome. @mkistler cool with that?

mkistler replied:
Static C functions are perfectly acceptable.

mkistler commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2798 @@
+    return prop.typeCode;
+}
+
+- (NSString *)destinationWithName:(NSString *)name withEntityName:(NSString *)ent
```
Should this method be declared in the interface?

jimix replied:
ACK, I wonder why I can call it?

mkistler commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2798 @@
+    return prop.destination;
+}
+
+- (NSString *)xformWithName:(NSString *)name withEntityName:(NSString *)ent
```
Should this method be declared in the interface?

jimix replied:
ACK

mkistler commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2798 @@
+{
+    if (![[self class] isEqual:[CDTIncrementalStore class]]) {
+        return;
+    }
```
Nice!  Very good idea to protect this code from running more than once.

mkistler commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2798 @@
+     *  - CDTREPLICATION_LOG_CONTEXT
+     *
+     *  We are interested in:
+     *  - CDTTD_REMOTE_REQUEST_CONTEXT
```
And CTDTDOCUMENT_REVISION_LOG_CONTEXT?

jimix replied:
ACK

mkistler commented on Classes/common/CDTIncrementalStore.h:
```diff
@@ -0,0 +1,121 @@
+#import <CloudantSync.h>
+
+extern NSString *const kCDTISErrorDomain;
+extern NSString *const kCDTISException;
```
I recommend dropping the 'k' prefix on these constant names.  Neither Apple nor Cloudant uses the 'k' convention (Cloudant's error domain is CDTDatastoreErrorDomain).

jimix replied:
ACK... probably :+1: 
I actually switched back and forth based on my experience with existing code.
Both cloudant (see `CDTIndexManager.[hm]`) and Cocoa use it a lot.
It is not mentioned in the Apple [Coding Guidelines for Cocoa](https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingIvarsAndTypes.html).
Apparently it is Hungarian Notation.
Personally, I don't care as long as it's consistent and follows some prescribed style guide and since they quote the cocoa style.. I shall remove them.

mkistler commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2798 @@
+                num = @(0);  // not sure what to do here
+                dbl = 0.;
+                meta[kCDTISFPNaNKey] = @"true";
+            }
```
Why not just capture the full 64-bit value as a hex string?  This would give you a single means to capture plus/minus infinity and NAN (in all it's various forms!).

jimix replied:
There is a brief discussion on this [here](https://github.com/jimix/CDTDatastore/pull/18/files#diff-a19a528c6ca4f3a8a6d02158a1eaf4d8R498)
So there are really three goals here:
1. provide a JSON compatible `Number` for `NaN` et. al. that can be useful for expressions
   - using `0` for `NaN` might be dubious here, but I can't think of a better value
2. detect `NaN` et. al. without needing to know IEEE-754
3. use IEEE-754 as an absolute store

Having said that, two things I've discovered recently:
1. using `Booleans` with three object keys is cumbersome, better to have a string object with three possible values.
   - I will likely change this soon.
2. If a non-IEEE-754 capable client updates the JSON value, how can it update the image?
   - don't have anything in mind for this

jimix replied:
making this issue #22

mkistler commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2798 @@
+ *
+ *  @return YES/NO
+ */
+- (BOOL)deleteManagedObject:(NSManagedObject *)mo error:(NSError **)error
```
Does CoreData handle cascading deletes above the incremental store layer?  You implementation of delete seems to depend on this.  It might be worth adding a note in the block comment on this assumption/dependency.

jimix replied:
ACK,
Yes I expect (and recall witnessing) coreData handling the cascade.
I'll note it.

mkistler commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2798 @@
+}
+
+/**
+ *  Create a dictionary of values for a the attributes of a Managed Object from
```
Typo: "a the" -> "the"

mkistler commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2798 @@
+    return [NSDictionary dictionaryWithDictionary:values];
+}
+
+- (BOOL)commWithRemote:(CDTReplicator *)rep
```
It seems that you have embedded the replicator functions into the incremental store.  There is another approach that we should at least consider: exposing the local Cloudant-sync db and allow the application to manage replication.

jimix replied:
making this a discussion point #23

jimix committed:
Add a version to the Core Data model that can be easily migrated to.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixup: update oops instructions

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Initial migration test

Does not actually test replication but I think it is sufficient to
perform the test locally.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Fixes #4: Support Migration of Managed Objects

This patch includes:
- Only store Entity hashes, and no other apple specific data (default on)
- Update the Stored Object Model if the Managed Object Model has changed (default off)
- Check for Entity version mismatches (default off)
- ObjectModel now provides a version hash dictionary that can be used to compare against the one in metatdata

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
CR: spellcheck pass

jimix committed:
CR: adopt correct "initializer pattern"

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
CR: use `instancetype` for initializers and add missing prototypes

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
CR: update comment on logging

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
CR: k=constant is no longet part of the Cocoa Coding Guidlines

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
CR: add comment on deleteManagedObject as to cascading deletes

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix commented on Classes/common/CDTIncrementalStore.h:
```diff
@@ -0,0 +1,121 @@
+ *
+ *  @return NSURL
+ */
++ (NSURL *)localDir;
```
How to differentiate between different IS stores in the same PSC

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2948 @@
+ *  @return Encoded array
+ */
+- (NSDictionary *)encodeAttribute:(NSAttributeDescription *)attribute
+                       withObject:(id)obj
```
Consider "attributeValue"

jimix replied:
ACK

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2948 @@
+                name : str,
+            };
+        }
+        case NSBooleanAttributeType:
```
Can I express this as a literal true?

jimix replied:
NACK, Not actually an issue, see #25

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2948 @@
+
+            if ([dec isEqual:[NSDecimalNumber notANumber]]) {
+                meta[CDTISFPNaNKey] = @"true";
+                desc = @"0";
```
Use the absence of a the `desc` for NaN
Track as #26

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2948 @@
+        } else if ([prop isKindOfClass:[NSRelationshipDescription class]]) {
+            NSRelationshipDescription *rel = prop;
+            enc = [self encodeRelation:rel withObject:obj error:&err];
+        } else {
```
have an `else if` clause for the third property type
Tracking with #27

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2948 @@
+    }
+
+    // just checking
+    NSArray *entitySubs = [[mo entity] subentities];
```
This debug code could be behind a switch

jimix replied:
ACK

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2948 @@
+                NSManagedObjectID *moid =
+                    [self decodeRelationFromEntityName:entityName withRef:ref withContext:context];
+                // we cannot return nil
+                if (!moid) {
```
Check this line.. do you really need it?

jimix replied:
Add comment on why you need to do this.

jimix replied:
ACK

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2948 @@
+    // do the actual attributes first
+    newRev.docId = docID;
+    newRev.body = [self propertiesFromManagedObject:mo withBlobStore:blobStore];
+    newRev.body[CDTISObjectVersionKey] = @"1";
```
Code review made it this far

jimix committed:
GCR: Nits

jimix committed:
Fixes #24: Export databaseName so individual object may be identified

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Fixes #26: Use Null to represent NaN instead of zero

This will ensure the value will not be evaluated in a predicate

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Test NaN in predicate fetches

Test for Issue #26

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Fixes #22: merge Non-Finite (Special) values as a sible enum property

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Splitsville: CDTIncrementalStore.m is so big that github refuses to display it

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2398 @@
+        [upRev.attachments addEntriesFromDictionary:blobStore];
+    }
+
+    // delete all changed properties, in case they are being removed.
```
@pbohrer asks:

> When can a property be deleted?

jimix replied:
**_YES**_ we can skip the removal of the objects.
Early in development, we sprinkled the meta-data around the document body, which could cause a stale meta-data entry to be present and stale.  This is no longer the case, and properties requiring meta always have a collected meta-data dictionary associated with them.  This results in a consistent document body that requires no clean up.
A comment will be added that explains this.

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2398 @@
+        return NO;
+    }
+
+    if (self.progressBlock) {
```
@pbohrer says:

> This is not sufficient to test if a "sync" is in progress, since the block provided by the interface could be `nil`
> Some form of concurrency  "guard" is necessary here

jimix replied:
@mkistler says:

> Suggest a GCD/NSOperation solution.  This may enable some sync/conflict mechanism to be provided once the `optLock` stuff is in place.

jimix replied:
Group Code review made it to here

jimix replied:
This will go away with the new sync work tracked by Issues #23 and #21

jimix committed:
GCR: No need to delete modified properties

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2409 @@
+
+- (BOOL)pushToRemote:(NSError **)error
+        withProgress:
+            (void (^)(BOOL last, NSInteger processed, NSInteger total, NSError *err))progress;
```
These `;` at the end of declarations should be removed.

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2409 @@
+    self.puller = nil;
+}
+
+static NSString *fixupName(NSString *name)
```
We should always check compliance and report the failure adivising the switch.
Should doc this as well.

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2409 @@
+    return [NSDictionary dictionaryWithDictionary:metaData];
+}
+
+- (NSDictionary *)updateObjectModel
```
Does implicit migration de-dup the older objects? @pbohrer recalls no duplicates.

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2409 @@
+
+    CDTDocumentRevision *rev = [self.datastore getDocumentWithId:docID error:&err];
+    if (!rev) {
+        self.run = @"1";
```
Get rid of self.run.. the original debug code is not worth it.

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2409 @@
+ *
+ *  @return YES/NO
+ */
+- (BOOL)checkMetaData:(NSDictionary *)metaData error:(NSError **)error
```
This function is not useful

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2409 @@
+}
+
+#pragma mark - Database Delegates
+/**
```
Move other functions above to here

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2409 @@
+}
+
+/**
+ * Called when a state transition to ERROR is completed.
```
more comments what methods are being called by what object.

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2409 @@
+    NSArray *sds = [fetchRequest sortDescriptors];
+    if (sds.count) {
+        if (sds.count > 1) {
+            oops(@"not sure what to do here");
```
multiple sort descriptors should be reasonable to support.
should add a test for it so we can remove this oops

Tracking as Issue #29

jimix replied:
@mkistler asks, does coundant sync limit requests to 100 results like the cloudantDB does.
There is indeed core data support for a fetch/limit/window. should tie these together.

jimix replied:
Answer.. no, it imposes no limits.

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2409 @@
+ *  The queries currently supported by the backing store are:
+ *  * `{index: @{@"max": value}}`: index <= value
+ *  * `{index: value}`: index == value
+ *  * `{index: @{@"min": value}}`: index >= value
```
Suspect that this does not work for double.

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2409 @@
+                [NSException raise:CDTISException format:@"Can't do substring matches: %@", value];
+                break;
+            }
+            // FIXME? I hope this deals with collections
```
This has been tested, remove the fixme

jimix commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -0,0 +1,2409 @@
+        case NSNotEqualToPredicateOperatorType:
+        case NSGreaterThanPredicateOperatorType:
+        case NSLessThanPredicateOperatorType:
+        case NSMatchesPredi
```
Please expand on this

jimix replied:
comment on how this is for dedup support.. add url to the apple docs. This simulates SQL sections

jimix replied:
This is really application error, make this an exception.
Describe the exception better, "stale fetch request"

jimix replied:
You should fill in toMany at decode time. This can be simplified

jimix commented on ReplicationAcceptance/ReplicationAcceptance/ReplicatorCoreData.m:
```diff
@@ -0,0 +1,571 @@
+                                                                   destinationModel:toMom
+                                                                              error:&err];
+        XCTAssertNotNil(mapMom, @"Failed to create mapping model");
+        XCTAssertNil(err, "Error: %@", err);
```
`@""`

jimix committed:
Add a test for AsyncFetchRequests

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Fixes #28: Migration of databases dies when updating metadata

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
GCR: Get rid of self.run.. the original debug code is not worth it.

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
GCR: NIT: remove FIXME

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
GCR: missing @ in front of some string literals

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Use CloudantSync interfaces for pushing and pulling

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

Commit:
XCTest for CoreData batch update request (Issue #30).

Commit:
A basic implementation that achieves the behavior expected in the initial test.

But there is clearly more work to be done here ... starting with more tests.

Commit:
Another test and an improved implementation.

jimix committed:
Allow arbitrary number of secondary remotes

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Use new push/pull interfaces in tests

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

Commit:
Initial conversion to CloudantQuery.  All existing tests passing.

Commit:
More conservative approach to indexing.

Commit:
Test and minor fixes for multiple sort descriptors.

Commit:
Add some predicate tests and fix handling of not-equal predicates.

Commit:
Add a test for negation -- currently failing (bug in CloudantSync?).

Commit:
Fix stuff from code review.

Commit:
Cannot support NOT compound expressions (for now).

Commit:
Recode one of the predicate tests to verify results using filteredArrayWithPredicate.

Commit:
Add JimiSwitch to control support for batch updates.

rhyshort committed:
Add CDTDatastore tests to build script

Added ability to run tests from the CDTDatastore workspace, in addition
to replication acceptance tests.

ReplicationAcceptance tests are only run when -nightly true is passed
into the build script. if -ra false is passed in or nothing is
passed in the build script will only run the tests from CDTDatastore

JUnit result files are located in build/reports in the format <platform>_<unit|ra>.xml

mikerhodes committed:
Support cancelling of CDTFetchChanges

We now periodically check for the NSOperation's cancelled
property and bail early rather than always running the whole
changes fetch process.

BugId: 46038

Commit:
The straightforward changes from code review.

mikerhodes committed:
Fix calling completion block after cancel edge case

Commit:
Batch update should modify the underlying document rather than the managed object.

Commit:
A slightly better hack to allow creation of NSBatchUpdateResults.

Commit:
Add support and tests for predicateWithValue type predicates.

Commit:
Fix message in exception for unsupported predicate and add test.

Commit:
Fix missing index problem for predicates on expressionForEvaluatedObject.

Commit:
Small hack to support IncrementalStore tests with sql, to validate tests.
Also munge value(s) for predicates on NSEvaluatedObjectExpressionType.

Commit:
Straightforward fixes from Jimix github comments.

Commit:
Finish rebase by moving batchUpdate to use new query functions.

Commit:
Merge in Jimi's suggested changes to do the update directly on the document.

Commit:
Need to set stalenessInterval to get refresh to really refresh.

jimix committed:
Verify and deal with SQLite failures for CDTIS tests

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

rhyshort committed:
Merge pull request #124 from cloudant/run_all_ios_tests

Add CDTDatastore tests to build script

alfinkel committed:
Add Index class for index attribute validation

- Add CDTQIndex.h/.m and add field names, index name, index type, and
index settings as attributes.
- Provide attribute validation within convenience constructor method
calls.
- Provide utility methods to compare an index's type and settings
with existing values and to retrieve index settings as an NSString.
These methods will be used by the index creator as part of its index
creation process (in a separate commit).
- Add appropriate tests.

jimix committed:
Code Review Comments

This patch contains fixes for:
- Comment typos
- More updates to documentation

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

rhyshort committed:
Update import statements to be compatible with frameworks

When compiling CDTDatastore with the `use_frameworks!` cocoapods
directive depedancies are built as frameworks. The header import
statements for frameworks need to be in the form
<framework/header.h> and are updated to follow said form.

Note: CDTDatastore will not compile under the `user_frameworks!`
cocoapods directive until FMDB supports being built as a framework.

BugzID: 46367

rhyshort committed:
Add missing copyright

Add missing copyright header to CDTLogging.h

jimix committed:
Allocate a replication factory at the beginning and reuse it

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

mikerhodes committed:
Check cancelled before reading changed docs from disk

I only check before reading the documents from disk rather than
putting them into the updatedRevisions array as that loop does
no externally visible work (unlike the blocking read from the
database).

mikerhodes committed:
Replace "request" with "operation" in docs

This implies an operation object can only be used for
one set of requests, whereas using request was ambiguous
on this point.

mikerhodes committed:
Use 'fetch operation' not 'fetch object' in docs

mikerhodes committed:
Once -main is called, ignore property changes

mikerhodes committed:
No need to copy blocks manually, property is copy

jimix committed:
Split getting values from persistent store directly from document

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
fixup: cleanup graphviz object so the incremental store doesn't need to leak stuff

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

rhyshort committed:
Merge pull request #131 from cloudant/46367_support_frameworks

46367 support frameworks

mikerhodes committed:
Merge pull request #125 from cloudant/46038-cancellable-fetchchanges

Support cancelling of CDTFetchChanges

rhyshort committed:
Check DatastoreManager passed to CDTReplicatorFactory

Check if the DatastoreManager passed to CDTReplicatorFactory
is nil, if so CDTReplicatoryFacotry initWithDatastoreManager
returns nil.

BugzID 46518

Fixes #132

alfinkel committed:
Merge pull request #130 from cloudant/46368-fts-add-index-class

Merged

jimix committed:
Our own Replicator that can manage conflicts

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
RA tests for merge conflicts

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

rhyshort committed:
Merge pull request #135 from cloudant/46518_nil_datastore_replicator

Check DatastoreManager passed to CDTReplicatorFactory

jimix committed:
update document for replictaion and conflicts

Signed-off-by: Jimi Xenidis <jimix@pobox.com>
